lib/src/main/kotlin/StateNode.kt

```diff
@@ -1,17 +1,35 @@
-internal class StateNode<TAction, TState>(
-        parent: NodeBase?,
+class StateNode<TState, TAction>(
+        parent: StateNode<TState, TAction>?,
+        inducingAction: TAction?,
         val state: TState,
-        val validActions: List<TAction>,
-        val isTerminal: Boolean) : Node<ActionNode<TAction, TState>>(parent) {
+        val validActions: Collection<TAction>,
+        val isTerminal: Boolean) : Node<TAction, StateNode<TState, TAction>>(parent, inducingAction) {
 
-    fun parentAction() : ActionNode<TAction, TState>? {
-        return if (parent == null)
-            null
-        else
-            parent as ActionNode<TAction, TState>
+    private val children = mutableMapOf<TAction, MutableCollection<StateNode<TState, TAction>>>()
+
+    override fun addChild(child: StateNode<TState, TAction>) {
+        if (child.inducingAction == null) {
+            throw Exception("Inducing action must be set on child")
```

Comment:
How would this error actually arise, could we prevent this from occuring?

Reply:
You can new up a Node with null for parent and inducing action. That is how root nodes are represented. This check guards against those types of invalid nodes.